File: app/src/main/res/drawable/viewmodell/LoginViewModel.java
package com.example.mvvm_practice.viewmodell;

import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.databinding.BaseObservable;
import androidx.databinding.Bindable;
import androidx.databinding.ObservableField;

import com.example.mvvm_practice.BR;
import com.example.mvvm_practice.model.ReqToken;
import com.example.mvvm_practice.model.User;
import com.example.mvvm_practice.network.ApiService;
import com.example.mvvm_practice.network.OnCallBack;
import com.example.mvvm_practice.view.MainActivity;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class LoginViewModel extends BaseObservable {
    public OnCallBack callBack;

    public void setCallBack(OnCallBack callBack) {
        this.callBack = callBack;
    }

    public ObservableField<String> username = new ObservableField<>();
    public ObservableField<String> password = new ObservableField<>();

    public void onClickLogin(String user){
        getToken();
//        if(mlistUser == null || mlistUser.isEmpty()){
//            return;
//        }
//        boolean isUserInCom = false;
//        for(User user : mlistUser){
//            if(email.equals(user.getUserEmail()) && password.equals(user.getUserPassword())){
//               isUserInCom = true;
//               break;
//            }
//        }
//        if (isUserInCom){
//           Log.i(TAG1, "");
//        }
//        else {
//            Log.i(TAG, "");
//        }
    }

    private void sendPost(String username, String pass, String token) {
        User user = new User(username, pass, token);
        ApiService.apiService.verifyLogin(user).enqueue(new Callback<ReqToken>() {
            @Override
            public void onResponse(Call<ReqToken> call, Response<ReqToken> response) {
                if (response.code() == 200 || response.code() == 201){
                    Log.e("abc", response.body() + " "+ response.code());
                    callBack.loginSuccess();
                } else {
                    Log.e("abc", response.errorBody() + "");
                }
            }

            @Override
            public void onFailure(Call<ReqToken> call, Throwable t) {

            }
        });
    }

    private void getToken(){
        ApiService.apiService.getToken()
                .enqueue(new Callback<ReqToken>() {
                    @Override
                    public void onResponse(@NonNull Call<ReqToken> call,@NonNull Response<ReqToken> response) {
                        Log.e("abc", response.body() + " "+ response.code());
                        sendPost(username.get(), password.get(), (response.body()).toString());
                    }

                    @Override
                    public void onFailure(Call<ReqToken> call, Throwable t) {
                        Log.i("Errors",t.getMessage());
                    }
                });
    }
}

File: app/src/main/java/com/example/mvvm_practice_kotlin/viewmodell/LoginViewModel.kt
package com.example.mvvm_practice_kotlin.viewmodell

import android.content.Context
import android.util.Log
import androidx.databinding.BaseObservable
import androidx.databinding.ObservableField
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.mvvm_practice_kotlin.model.ReqToken
import com.example.mvvm_practice_kotlin.model.User
import com.example.mvvm_practice_kotlin.network.ApiService
import com.example.mvvm_practice_kotlin.repository.AppRepository
import com.example.mvvm_practice_kotlin.utils.OnCallBack
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import javax.inject.Inject

@HiltViewModel
class LoginViewModel @Inject constructor(private val repository: AppRepository): ViewModel() {

    val user = MutableLiveData<String>()
    val passw = MutableLiveData<String>()
    var userName = ""
    var password = ""

    val tokenResponse = MutableLiveData<ReqToken?>()

    fun getToken(){
        viewModelScope.launch {
            repository.getToken().let {
                if (it.isSuccessful){
                    tokenResponse.value = it.body()
                } else {
                    println("Co loi xay ra")
                }
            }
        }
    }


     fun login(token: String) = viewModelScope.launch {
        val gettoken = repository.login(userName, password, token).let {response ->
               if(response.isSuccessful){
                   user.postValue(response.body()?.requestToken) // gui reponse den user thong qua postValue
                   println("==========Login OK")
               }else{
                   Log.d("123","Error get token")
               }
           }
         }

    fun sendPost(token : String){

      }
    }

//    private val _userName2 = MutableLiveData<String>()
//    var userName2: LiveData<String> = _userName2


//    private var _userName0: MutableStateFlow<String> = MutableStateFlow("")
//    var userName: StateFlow<String> = _userName0
//
//    private var _password0: MutableStateFlow<String> = MutableStateFlow("")
//    var password: StateFlow<String> = _password0

//    private fun sendPost( token: String) {
//        val user = User("nightowl990", "123456", token)
//        ApiService.apiService.verifyLogin(user).enqueue(object : Callback<ReqToken?>{
//            override fun onResponse(call: Call<ReqToken?>, response: Response<ReqToken?>){
//                if (response.code() == 200 || response.code() == 201) {
//                    Log.e("success", response.body().toString() + " " + response.code())
//                    callBack.loginSuccess()
//                } else {
//                    Log.e("fail", response.errorBody().toString())
//                }
//            }
//
//            override fun onFailure(call: Call<ReqToken?>, t: Throwable?) {
//            }
//        })
//    }

//    private fun getToken(){
//        ApiService.apiService.getToken().enqueue(object : Callback<ReqToken> {
//            override fun onResponse(call: Call<ReqToken>, response: Response<ReqToken>) {
//                if (response.code() == 200 || response.code() == 201){
//                    Log.e("abc", response.body().toString() + " " + response.code())
//                    sendPost(((response.body()) as ReqToken).toString())
//                } else{
//                    Log.e("abc",   "---------------${response.code()}")
//                }
//            }
//
//            override fun onFailure(call: Call<ReqToken?>?, t: Throwable) {
//                Log.e("Errors", t.message!!)
//            }
//        })
//    }




